src/lib/support/PersistentArray.h

```diff
@@ -0,0 +1,316 @@
+     *        is used to identify the matching entry within the table.
+     * @param index Zero-based position within the table.
+     * @param entry Used as an input for comparison, and output for the rest of the values.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if no entry matches the given input.
+     */
+    CHIP_ERROR Get(EntryType & value, bool do_create = false)
+    {
+        CHIP_ERROR err = this->Load();
+        VerifyOrReturnError(CHIP_ERROR_NOT_FOUND == err || CHIP_NO_ERROR == err, err);
+
+        // Check existing
+        for (size_t i = 0; i < this->mCount; ++i)
+        {
+            EntryType & e = mEntries[i];
+            if (this->Compare(value, e))
+            {
+                ReturnErrorOnFailure(this->Copy(value, e));
+                return CHIP_NO_ERROR;
+            }
+        }
+        // Insert last ?
+        VerifyOrReturnError(do_create, CHIP_ERROR_NOT_FOUND);
+        VerifyOrReturnError(this->mCount < this->mLimit, CHIP_ERROR_INVALID_LIST_LENGTH);
+        ReturnErrorOnFailure(this->Copy(mEntries[this->mCount++], value));
+        return this->Save();
+    }
+
+    /**
+     * @brief Removes the stored entry at a given index. Remaining entries are
+     *        shifted one position down.
+     * @param index Zero-based position within the table.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if index is greater than the index of the last entry on the table.
+     */
+    CHIP_ERROR Remove(size_t index)
+    {
+        ReturnErrorOnFailure(this->Load());
+        VerifyOrReturnError(index < this->mLimit, CHIP_ERROR_NOT_FOUND);
+        VerifyOrReturnError(index < this->mCount, CHIP_ERROR_NOT_FOUND);
+        EntryType old = mEntries[index];
+        this->mCount--;
+        for (size_t i = index; i < this->mCount; ++i)
+        {
+            mEntries[i] = mEntries[i + 1];
+        }
+        EntryType & last = mEntries[this->mCount];
+        Clear(last);
+        ReturnErrorOnFailure(this->Save());
+        OnEntryRemoved(old);
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Removes the value matching the given entry. The Compare() method
+     *        is used to identify the matching entry within the table.
+     * @param entry Identifies the value to be removed.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if no entry matches the given input.
+     */
+    CHIP_ERROR Remove(EntryType & value)
+    {
+        ReturnErrorOnFailure(this->Load());
+        for (size_t i = 0; i < this->mCount; ++i)
+        {
+            if (this->Compare(mEntries[i], value))
+            {
+                return Remove(i);
+            }
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Removes all the entries in the table.
+     * @return CHIP_NO_ERROR on success.
+     */
+    CHIP_ERROR RemoveAll()
+    {
+        for (uint16_t i = 0; i < mCount; ++i)
```

Comment:
Don't we need to Load() somewhere here?  Otherwise mCount is not going to be set...  This should be covered by some sort of test, I would think.

Reply:
We don't. There is a Load() on each individual Remove(). I have added a flag to prevent unnecessary loads on subsequents calls.
That said, it doesn't hurt, and it is clearer to put an explicit Load() before the loop, so I have.

Reply:
> There is a Load() on each individual Remove()

That does not help if we are not loaded and mCount is 0, right?  So we really do need the load up front.

Reply:
If not loaded, then it would load on first Remove. After that, if mCount is 0, then there is nothing to remove.
But again, I agree adding the Load here is the way to go.